### PR TITLE
Updates to Parkinson to Synchronise With Open Source with App Store Version

### PR DESCRIPTION
**Pull request by PacificCoaster**
This Pull Request contains four primary commits:

o    Changes in Tasks and Schedules JSON file for New Scheduler

o    Changes in Survey JSON file keywords

o    Changes to Parkinson to cater to new Pre-Survey capability, where the most recent 'Did you take your medications' result is stashed away so that the result can be uploaded even when the question is not asked.

o    Plus a Gallimaufry of minor typographic changes . . .

### CODE REVIEW

**Comment by PacificCoaster** [comment]
'xxxxx' should have read 'four'

File: Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+- (void)viewDidLoad
+{
+    [super viewDidLoad];
+}
+
+- (void)didReceiveMemoryWarning
+{
+    [super didReceiveMemoryWarning];
+}
```
Comment (YuanZhu-apple) [inline Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m]: Please remove these two overrides if they are not used.

File: Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+{
+    if (reason == ORKTaskViewControllerFinishReasonCompleted) {
+        BOOL  foundSurveyStep = NO;
+        
+        for (ORKStepResult *stepResult in self.result.results) {
+            if ([stepResult.identifier isEqualToString:kMomentInDayStepIdentifier] == YES) {
+                foundSurveyStep = YES;
+                break;
+            }
+        }
```
Comment (YuanZhu-apple) [inline Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m]: `ORKCollectionResult` has a convenient method `resultForIdentifier:`.

File: Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+                                                  kMomentInDayChoiceAnswerKey : result.choiceAnswers.lastObject,
+                                                  kMomentInDayQuestionTypeKey : @(result.questionType),
+                                                  kMomentInDayIdentifierKey   : result.identifier,
+                                                  kMomentInDayStartDateKey    : result.startDate,
+                                                  kMomentInDayEndDateKey      : result.endDate,
+                                                  };
+                    NSUserDefaults  *defaults = [NSUserDefaults standardUserDefaults];
+                    [defaults setObject:dictionary forKey:kMomentInDayUserDefaultsKey];
+                    [defaults synchronize];
+                }
```
Comment (YuanZhu-apple) [inline Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m:181]: Why do the stash? Is a feature which is missing from ResearchKit?

File: Parkinson/Startup/APHAppDelegate.m [hunk 4]
```diff
@@ -716,40 +678,36 @@ - (ORKTaskViewController *)consentViewController
...
-                     forKey:kJsonTasksKey];
-    
-    [newDictionary setValue:newSchedulesArray
-                     forKey:kJsonSchedulesKey];
-    
+
+    [newDictionary setValue:[dictionary objectForKey:kJsonTasksKey] forKey:kJsonTasksKey];
+
+    [newDictionary setValue:newSchedulesArray forKey:kJsonSchedulesKey];
+
```
Comment (YuanZhu-apple) [inline Parkinson/Startup/APHAppDelegate.m]: Why return a mutable copy NSDictionary? 
If a copy is needed, why just do `return [dictionary copy]`

File: Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m (hunk 3 above)
Comment (PacificCoaster) [inline Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m:181]: > On Jun 29, 2015, at 2:00 PM, Yuan Zhu notifications@github.com wrote:
> 
> In Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m https://github.com/ResearchKit/mPower/pull/6#discussion_r33512747:
> 
> > -                ORKChoiceQuestionResult  *result = (ORKChoiceQuestionResult *)object;
> > -                    //
> > -                    //  stash the object away
> > -                    //
> > -                if (result.choiceAnswers.lastObject != nil) {
> > -                    NSDictionary  *dictionary = @{
> > -                                                  kMomentInDayChoiceAnswerKey : result.choiceAnswers.lastObject,
> > -                                                  kMomentInDayQuestionTypeKey : @(result.questionType),
> > -                                                  kMomentInDayIdentifierKey   : result.identifier,
> > -                                                  kMomentInDayStartDateKey    : result.startDate,
> > -                                                  kMomentInDayEndDateKey      : result.endDate,
> > -                                                  };
> > -                    NSUserDefaults  *defaults = [NSUserDefaults standardUserDefaults];
> > -                    [defaults setObject:dictionary forKey:kMomentInDayUserDefaultsKey];
> > -                    [defaults synchronize];
> > -                }
> >   Why do the stash? Is a feature which is missing from ResearchKit?
> 
> —
> Reply to this email directly or view it on GitHub https://github.com/ResearchKit/mPower/pull/6/files#r33512747.

The stash is done as a special case introduced at a late stage in the four Parkinson activities.

The idea is to inject a step at the beginning of the activity to ask the patient if they have taken their
medication prior to doing the  activity.   If they then do another activity within a time limit (currently
twenty minutes), they won’t get pestered again.

The original flow was that the results of that question would be uploaded with the other results.

Then the researchers asked us to upload the results even when the question didn’t get asked, so
we had to stash away the most recent results to be uploaded.

We could debate on whether ResearchKit should have that capability folded in.  The requirement
seems pretty esoteric.    Course, if other researchers ask for the same capability, that could be
the impetus to create the capability in a more formal way.

```
Best Wishes,
    . . . . . . . .    Henry
```

File: Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m (hunk 2 above)
Comment (YuanZhu-apple) [inline Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m]: Thanks for explaining.

Can you use `NSSecureCoding` to save the `stepResult` into NSUserDefaults? 
So you don't need to rebuild the object on restoring.

File: Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m (hunk 3 above)
Comment (YuanZhu-apple) [inline Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m:181]: Thanks for explaining.

Can you use `NSSecureCoding` to save the `stepResult` into NSUserDefaults? 
So you don't need to rebuild the object on restoring.

File: Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m (hunk 2 above)
Comment (YuanZhu-apple) [inline Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m]: `ORKTaskResult` is a subclass `ORKCollectionResult`

File: Parkinson/Startup/APHAppDelegate.m (hunk 4 above)
Comment (PacificCoaster) [inline Parkinson/Startup/APHAppDelegate.m]: > On Jun 29, 2015, at 2:13 PM, Yuan Zhu notifications@github.com wrote:
> 
> In Parkinson/Startup/APHAppDelegate.m https://github.com/ResearchKit/mPower/pull/6#discussion_r33514222:
> 
> > ```
> >  for (NSDictionary *schedule in schedules) {
> >      [newSchedulesArray addObject:schedule];
> >  }
> > ```
> > -   
> > -    [newDictionary setValue:[dictionary objectForKey:kJsonTasksKey]
> > 
> > ## \-                     forKey:kJsonTasksKey];
> > -    [newDictionary setValue:newSchedulesArray
> > 
> > ## \-                     forKey:kJsonSchedulesKey];
> > 
> > +
> > -    [newDictionary setValue:[dictionary objectForKey:kJsonTasksKey] forKey:kJsonTasksKey];
> >   +
> > -    [newDictionary setValue:newSchedulesArray forKey:kJsonSchedulesKey];
> >   +
> >   Why return a mutable copy NSDictionary? 
> >   If a copy is needed, why just do return [dictionary copy]
> 
> —
> Reply to this email directly or view it on GitHub https://github.com/ResearchKit/mPower/pull/6/files#r33514222.

Entire method was redundant.   Deleted.

File: Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m (hunk 2 above)
Comment (PacificCoaster) [inline Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m]: > On Jun 29, 2015, at 2:45 PM, Yuan Zhu notifications@github.com wrote:
> 
> In Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m https://github.com/ResearchKit/mPower/pull/6#discussion_r33517507:
> 
> > -    }
> >   +}
> >   +
> >   +#pragma  mark  -  Step View Controller Delegate Methods
> >   +
> >   +- (void)taskViewController:(ORKTaskViewController *)taskViewController didFinishWithReason:(ORKTaskViewControllerFinishReason)reason error:(nullable NSError *)error
> >   +{
> > -    if (reason == ORKTaskViewControllerFinishReasonCompleted) {
> > -        BOOL  foundSurveyStep = NO;
> > -   
> > -        for (ORKStepResult *stepResult in self.result.results) {
> > -            if ([stepResult.identifier isEqualToString:kMomentInDayStepIdentifier] == YES) {
> > -                foundSurveyStep = YES;
> > -                break;
> > -            }
> > -        }
> >   ORKTaskResult is a subclass ORKCollectionResult
> 
> —
> Reply to this email directly or view it on GitHub https://github.com/ResearchKit/mPower/pull/6/files#r33517507.

Okay, did that.  Thank you for the reminder—I’d forgotten all about that method.

File: Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m (hunk 3 above)
Comment (PacificCoaster) [inline Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m:181]: > On Jun 29, 2015, at 2:43 PM, Yuan Zhu notifications@github.com wrote:
> 
> In Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m https://github.com/ResearchKit/mPower/pull/6#discussion_r33517299:
> 
> > -                ORKChoiceQuestionResult  *result = (ORKChoiceQuestionResult *)object;
> > -                    //
> > -                    //  stash the object away
> > -                    //
> > -                if (result.choiceAnswers.lastObject != nil) {
> > -                    NSDictionary  *dictionary = @{
> > -                                                  kMomentInDayChoiceAnswerKey : result.choiceAnswers.lastObject,
> > -                                                  kMomentInDayQuestionTypeKey : @(result.questionType),
> > -                                                  kMomentInDayIdentifierKey   : result.identifier,
> > -                                                  kMomentInDayStartDateKey    : result.startDate,
> > -                                                  kMomentInDayEndDateKey      : result.endDate,
> > -                                                  };
> > -                    NSUserDefaults  *defaults = [NSUserDefaults standardUserDefaults];
> > -                    [defaults setObject:dictionary forKey:kMomentInDayUserDefaultsKey];
> > -                    [defaults synchronize];
> > -                }
> >   Thanks for explaining.
> 
> Can you use NSSecureCoding to save the stepResult into NSUserDefaults? 
> So you don't need to rebuild the object on restoring.
> 
> —
> Reply to this email directly or view it on GitHub https://github.com/ResearchKit/mPower/pull/6/files#r33517299.
> 
> In principle, yes, we could replace that result —> dictionary and dictionary —> result.
> Presumably you are suggesting creating an archive and stashing that in NSUserDefaults.

However, the code has been in shipping versions of the App for a couple of months now,
and so we would be faced with a migration path if we wanted to switch to the NSSecureCoding
version. 

```
Best Wishes,
    . . . . . . . .    Henry
```

Comment (YuanZhu-apple) [inline Parkinson/TasksAndSteps/APHParkinsonActivityViewController.m:181]: OK, if this patch is from shipped app.

**Comment by PacificCoaster** [comment]
> On Jun 30, 2015, at 5:37 PM, Yuan Zhu notifications@github.com wrote:
> 
> Merged #6 https://github.com/ResearchKit/mPower/pull/6.
> 
> —
> Reply to this email directly or view it on GitHub https://github.com/ResearchKit/mPower/pull/6#event-344830239.
> 
> Many Thanks . . .

```
. . . . . . . .    Henry
```

**Pull request merged**